lib/matplotlib/axes/_axes.py

```diff
@@ -4241,9 +4238,9 @@ def _parse_scatter_color_args(c, edgecolors, kwargs, xshape, yshape,
         else:
             try:  # First, does 'c' look suitable for value-mapping?
                 c_array = np.asanyarray(c, dtype=float)
-                n_elem = c_array.shape[0]
-                if c_array.shape in [xshape, yshape]:
-                    c = np.ma.ravel(c_array)
+                n_elem = c_array.size
+                if n_elem == xsize:
```

Comment:
why do we define `n_elem` here?  its a confusing name now that its a size, and I don't see it adding anything.

Reply:
I didn't change this--n_elem is being used the same way it was before this PR.  Throughout, n_elem is the number of colors, regardless of whether they are coming from a mapping or from an rgba array.  It is used only in reporting an error, if one occurs.

Reply:
I extracted the code from `scatter` to make it more manageable. It's may not always be the best code style, but I didn't change too much to keep the extraction easier to review.

This should now be named `csize` in resemblence to `xsize`, `ysize`.